EIPS/eip-2200.md

```diff
@@ -0,0 +1,321 @@
+        0), remove `SSTORE_CLEARS_SCHEDULE` gas from refund
+        counter. We can prove that refund counter will never go below
+        0.
+      * If *new value* is 0 (also means that *current value* is not
+        0), add `SSTORE_CLEARS_SCHEDULE` gas to refund counter.
+    * If *original value* equals *new value* (this storage slot is
+      reset)
+      * If *original value* is 0, add `SSTORE_SET_GAS - SLOAD_GAS` to
+        refund counter.
+      * Otherwise, add `SSTORE_RESET_GAS - SLOAD_GAS` gas to refund
+        counter.
+
+An implementation should also note that with the above definition, if
+the implementation uses call-frame refund counter, the counter can go
+negative. If the implementation uses transaction-wise refund counter,
+the counter always stays positive.
+
+## Rationale
+
+This EIP mostly achieves what a transient storage tries to do
+(EIP-1087 and EIP-1153), but without the complexity of introducing the
+concept of "dirty maps", or an extra storage struct.
+
+* We don't suffer from the optimization limitation of
+  EIP-1087. EIP-1087 requires keeping a dirty map for storage changes,
+  and implicitly makes the assumption that a transaction's storage
+  changes are committed to the storage trie at the end of a
+  transaction. This works well for some implementations, but not for
+  others. After EIP-658, an efficient storage cache implementation
+  would probably use an in-memory trie (without RLP encoding/decoding)
+  or other immutable data structures to keep track of storage changes,
+  and only commit changes at the end of a block. For them, it is
+  possible to know a storage's original value and current value, but
+  it is not possible to iterate over all storage changes without
+  incurring additional memory or processing costs.
+* It never costs more gas compared with the current scheme.
+* It covers all usages for a transient storage. Clients that are easy
+  to implement EIP-1087 will also be easy to implement this
+  specification. Some other clients might require a little bit extra
+  refactoring on this. Nonetheless, no extra memory or processing cost
+  is needed on runtime.
+
+Regarding `SSTORE` gas cost and refunds, see Appendix for proofs of
+properties that this EIP satisfies.
+
+* For *absolute gas used* (that is, actual *gas used* minus *refund*),
+  this EIP is equivalent to EIP-1087 for all cases.
+* For one particular case, where a storage slot is changed, reset to
+  its original value, and then changed again, EIP-1283 would move more
+  gases to refund counter compared with EIP-1087.
+
+Examine examples provided in EIP-1087's Motivation (with `SLOAD_GAS` being
+`200`):
+
+* If a contract with empty storage sets slot 0 to 1, then back to 0,
+  it will be charged `20000 + 200 - 19800 = 400` gas.
+* A contract with empty storage that increments slot 0 5 times will be
+  charged `20000 + 5 * 200 = 21000` gas.
+* A balance transfer from account A to account B followed by a
```

Comment:
How does this affect the `CALLVALUE` opcode's costs, if at all? If it does not, why?

Reply:
I think this refers to the token balance transfer, because it is basically the storage modification operation inside one contract. Cost of CALLs is not affected.